Unit library implementation with Length, Area and Volume types

whaeck opened the pull request:
This is a first pull request to show how I have implemented the unit library and to check whether or not some of the choices I made are conform with what we want to do for NJOY21.

A first point of discussion is the organisation of the library. For this, I choose the utility library as a base. All classes that are to be used directly by the user are put in the unit namespace. As a result, each collection of unit types (e.g. LengthUnit, AreaUnit, VolumeUnit) and each physical quantity with an underlying unit (e.g. Length, Area, Volume) have their hpp directly in the unit directory. In order to "streamline" the various hpp files a bit more; I have also created a detail namespace. The boost units library allows a user to specify unit types in various ways: as a separate unit struct with an associated conversion factor to a base unit, as a scaled unit of a base unit, etc. By putting all of this in the detail namespace, I avoid giving the user of the unit library too much information to be able to work with the unit library.

For each physical quantity, I have defined an enumerator to represent the various units that will be implemented for the given quantity (e.g. unit::LengthUnit). For this enumerator, I also implemented iostream operators and conversion functions (this uses the xystream feature I introduced in the utility library). The underlying implementation uses two maps to associate string values with enum values. Because the maps are used in each of these functions, I have implemented them in a single cpp file - which is not the default way of doing things in NJOY21. The hpp file containing the enumerator declaration also contains the define statements for the unit classes associated to each unit (e.g. unit::Meter, unit::Centimeter, etc.) as well as the constants that can be used for construction of quantities in those units (e.g. unit::m, unit::cm, etc.).

The physical quantity itself is declared as a typedef using unit::Quantity in its own hpp file (e.g. unit::Lenght which is defined in unit::CentiMeter). Instead of creating multiple constructor methods, I choose to implement only one constructor method and multiple factory methods to create the physical quantities using its value and unit type (using a string or enum value).

For the moment, I have implemented the following physical quantities: length in cm, area in cm2 and volume in cm3.

jlconlin commented:
I took a brief look at this and have a few comments:

1. The tests pass—excellent
2. The file layout should be different. It doesn't follow the standard that we have developed. There isn't formal documentation for that (yet) so @whaeck couldn't have known how it should be done. I don't think it is a huge problem.
3. The compiler (clang) threw a bunch of warnings. Another (undocumented) policy is that we don't allow any compiler warnings. We'll have to figure out what the compiler warnings mean and fix the code to make the warnings go away.

whaeck commented on Area.hpp.in:
```diff
@@ -0,0 +1,72 @@
+   *  @brief An area value in cm2
+   *
+   *  The Area constructor allows a user to create area values in cm using
+   *  the cm unit as well as the other energy units (e.g. m2, mm2, um2, nm2 or
```
It should say area instead of energy

whaeck commented on Length.hpp.in:
```diff
@@ -0,0 +1,70 @@
+   *  @brief A length value in cm
+   *
+   *  The Length constructor allows a user to create length values in cm using
+   *  the cm unit as well as the other energy units (e.g. m, mm, um or nm).
```
It should say length instead of energy

whaeck commented:
I tried the initial implementation on the apple at LANL and I got a large number of compiler warnings related to boost. On g++ 4.9, I get no warning at all. I strive to write code that doesn't give compiler warnings but the clang warnings a will be difficult to get rid off (and I do not want to start modifying boost to remove them).

For the file layout, this is one of the reasons I did the pull request :-)
What file layout issues are you talking about? I tried copying the way it was done for other files in NJOY21 but I did notice that the layout tends to change from file to file

Concerning your fourth comment: YES. I do not know to set it up so I didn't do it (and I forgot to mention it in my initial comment).

apmccartney commented:
I apologize for the delay in reviewing this request. I'll review this pull in detail this pull Friday and have feedback posted before the end of the day.

whaeck commented:
No problem. We should take our time to do it right to make it fit into njoy21.

Can you also set up the travis continuous integration? I already added the travis.yml file (I added exclusions for boost in there already) but I do not know how to proceed from there.

whaeck commented on Area.hpp.in:
```diff
@@ -0,0 +1,72 @@
+   *    - unit::mm2
+   *    - unit::um2
+   *    - unit::nm2
+   *    - unit::barn
```
This was changed to unit::b. The comment wasn't changed.
